src/core-sdk-tasks/DownloadFile.cs

```diff
@@ -39,31 +54,57 @@ public override bool Execute()
                 File.Copy(filePath, DestinationPath);
+                return true;
             }
             else
             {
                 Log.LogMessage(MessageImportance.High, $"Downloading '{Uri}' to '{DestinationPath}'");
 
-                using (var httpClient = new HttpClient())
+                for (int retryNumber = 0; retryNumber < MaxRetries; retryNumber++)
                 {
-                    var getTask = httpClient.GetStreamAsync(Uri);
-
-                    try
+                    using (var httpClient = new HttpClient())
                     {
-                        using (var outStream = File.Create(DestinationPath))
+                        try
+                        {
+                            var httpResponse = await httpClient.GetAsync(Uri);
+
+                            if (httpResponse.StatusCode == HttpStatusCode.NotFound)
+                            {
+                                if (!string.IsNullOrEmpty(PrivateUri))
+                                {
+                                    httpResponse = await httpClient.GetAsync(PrivateUri);
+                                }
+
+                                // The Azure Storage REST API returns '400 - Bad Request' in some cases
+                                // where the resource is not found on the storage.
+                                // https://docs.microsoft.com/en-us/rest/api/storageservices/common-rest-api-error-codes
+                                if (httpResponse.StatusCode == HttpStatusCode.NotFound ||
+                                    httpResponse.ReasonPhrase.IndexOf("The requested URI does not represent any resource on the server.", StringComparison.OrdinalIgnoreCase) >= 0)
+                                {
+                                    return false;
+                                }
+                            }
+
+                            using (var outStream = File.Create(DestinationPath))
+                            {
+                                await httpResponse.Content.CopyToAsync(outStream);
+                            }
+
+                            return true;
+                        }
+                        catch (Exception e)
                         {
-                            getTask.Result.CopyTo(outStream);
+                            var secondaryUri = !string.IsNullOrWhiteSpace(PrivateUri) ? $"/'{PrivateUri}'" : string.Empty;
+                            Log.LogMessage($"Problems downloading file from '{Uri}'{secondaryUri}. {e.Message} {e.StackTrace}");
+                            File.Delete(DestinationPath);
                         }
                     }
-                    catch (Exception)
-                    {
-                        File.Delete(DestinationPath);
-                        throw;
-                    }
+
+                    Thread.Sleep(new Random().Next(3000, 10000));
```

Comment:
A few things:
* `await Task.Delay(...)` probably isn't any different for this particular PR, but would be more expected in an async method.
* Creating a new `Random` repeatedly can cause unexpected non-randomness due to seeding with current time (doesn't seem to matter in this case, but something to keep in mind).
* Instead of random, consider exponential falloff as a way to help better when the problem is resource contention. https://github.com/dotnet/arcade/blob/master/src/Common/ExponentialRetry.cs

Reply:
Thanks Davis. I'll adjust this.